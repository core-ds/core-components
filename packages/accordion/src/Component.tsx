import React, {
    AnchorHTMLAttributes,
    FC,
    KeyboardEvent,
    ReactNode,
    useCallback,
    useState,
} from 'react';
import cn from 'classnames';

import { Typography } from '@alfalab/core-components-typography';

import { DefaultControlIcon } from './components';
import { useRecalculateContentHeight } from './hooks';
import { ControlPosition } from './typings';

import styles from './index.module.css';

export type AccordionProps = {
    /**
     * Состояние компонента
     */
    expanded?: boolean;

    /**
     * Элемент заголовка
     */
    header: ReactNode;

    /**
     * Слот для элемента управления
     */
    control?: ReactNode;

    /**
     * Указывает компоненту - где будет размещен control
     */
    controlPosition?: ControlPosition;

    /**
     * Начальное состояние uncontrolled компонента
     */
    defaultExpanded?: boolean;

    /**
     * Основной элемент для отображения содержимого
     */
    children?: ReactNode;

    /**
     * Дополнительный класс обертки
     */
    className?: string;

    /**
     * Дополнительный класс для контейнера с заголовком
     */
    containerClassName?: string;

    /**
     * Дополнительный класс для header
     */
    headerClassName?: string;

    /**
     * Дополнительный класс для control
     */
    controlClassName?: string;

    /**
     * Дополнительный класс для body
     */
    bodyClassName?: string;

    /**
     * Обработчик смены состояний `expanded`
     */
    onExpandedChange?: (expanded: boolean) => void;

    /**
     * Идентификатор для систем автоматизированного тестирования
     */
    dataTestId?: string;
} & AnchorHTMLAttributes<HTMLDivElement>;

export const Accordion: FC<AccordionProps> = ({
    expanded,
    defaultExpanded = false,
    header,
    control,
    controlPosition = 'end',
    children,
    className,
    containerClassName,
    headerClassName,
    controlClassName,
    bodyClassName,
    onExpandedChange,
    dataTestId,
    ...rest
}) => {
    const uncontrolled = expanded === undefined;
    const [expandedState, setExpanded] = useState(uncontrolled ? defaultExpanded : expanded);
    const isExpanded = uncontrolled ? expandedState : expanded;

    const isStartPosition = controlPosition === 'start';

    const { contentRef, contentCaseRef } = useRecalculateContentHeight(isExpanded);

    const controlContent =
        control === undefined ? (
            <DefaultControlIcon expanded={isExpanded} startPosition={isStartPosition} />
        ) : (
            control
        );

    const headerContent =
        typeof header === 'string' ? (
            <Typography.Text view='primary-large' weight='medium'>
                {header}
            </Typography.Text>
        ) : (
            header
        );

    const bodyContent =
        typeof children === 'string' ? (
            <Typography.Text view='primary-medium'>{children}</Typography.Text>
        ) : (
            children
        );

    const handleExpandedChange = useCallback(() => {
        if (uncontrolled) {
            setExpanded(!isExpanded);
        }

        onExpandedChange?.(!isExpanded);
    }, [isExpanded, onExpandedChange, uncontrolled]);

    const handleKeyDown = (event: KeyboardEvent) => {
        if (event.key === 'Enter') {
            handleExpandedChange();
        }
    };

    return (
        <div {...rest} data-test-id={dataTestId} className={cn(styles.accordion, className)}>
            <div
                role='button'
                tabIndex={0}
                aria-expanded={isExpanded}
                onClick={handleExpandedChange}
                onKeyDown={handleKeyDown}
                className={cn(styles.container, containerClassName)}
            >
                <div
                    className={cn(styles.header, headerClassName, {
                        [styles.endPosition]: isStartPosition,
                    })}
                >
                    {headerContent}
                </div>

                <div
                    className={cn(styles.control, controlClassName, {
                        [styles.startPosition]: isStartPosition,
                    })}
                >
                    {controlContent}
                </div>
            </div>

            <div
                ref={contentRef}
                className={cn(styles.body, bodyClassName, { [styles.expandedBody]: isExpanded })}
            >
                <div className={cn(styles.bodyContent)} ref={contentCaseRef}>
                    {bodyContent}
                </div>
            </div>
        </div>
    );
};
