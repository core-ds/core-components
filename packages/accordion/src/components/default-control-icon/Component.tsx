import React, { FC } from 'react';
import cn from 'classnames';

import { ChevronDownMIcon } from '@alfalab/icons-glyph/ChevronDownMIcon';

import styles from './index.module.css';

type Props = {
    expanded: boolean;
    startPosition: boolean;
};
export const DefaultControlIcon: FC<Props> = ({ expanded, startPosition }) => (
    <ChevronDownMIcon
        className={cn(styles.icon, {
            [styles.expanded]: expanded,
            [styles.startPosition]: startPosition,
        })}
    />
);
