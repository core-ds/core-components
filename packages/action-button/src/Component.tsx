import React, {
    AnchorHTMLAttributes,
    ButtonHTMLAttributes,
    forwardRef,
    Fragment,
    useRef,
} from 'react';
import mergeRefs from 'react-merge-refs';
import cn from 'classnames';

import { Loader } from '@alfalab/core-components-loader';
import { useFocus } from '@alfalab/hooks';

import { useLoader } from './hooks';

import defaultColors from './default.module.css';
import styles from './index.module.css';
import invertedColors from './inverted.module.css';
import staticColors from './static.module.css';

const colorStyles = {
    default: defaultColors,
    inverted: invertedColors,
    static: staticColors,
};

/**
 * Минимальное время отображения лоадера - 500мс,
 * чтобы при быстрых ответах от сервера кнопка не «моргала».
 */
const LOADER_MIN_DISPLAY_INTERVAL = 500;

type Colors = 'default' | 'inverted' | 'static';

type ComponentProps = {
    /**
     * Иконка кнопки
     */
    icon: React.ReactNode;

    /**
     *  Размер кнопки
     */
    size?: 's';

    /**
     * Тип кнопки
     */
    view?: 'primary' | 'secondary';

    /**
     * Дополнительный класс
     */
    className?: string;

    /**
     * Дополнительный класс для обертки иконки
     */
    iconWrapperClassName?: string;

    /**
     * Значение href для ссылки
     */
    href?: string;

    /**
     * Заблокировать кнопку
     */
    disabled?: boolean;

    /**
     * Показать лоадер
     */
    loading?: boolean;

    /**
     * Id компонента для тестов
     */
    dataTestId?: string;

    /**
     * Палитра, в контексте которой используется кнопка
     */
    colors?: Colors;
};

type AnchorProps = ComponentProps & AnchorHTMLAttributes<HTMLAnchorElement>;
type ButtonProps = ComponentProps & ButtonHTMLAttributes<HTMLButtonElement>;
export type ActionButtonProps = Partial<AnchorProps | ButtonProps>;

export const ActionButton = forwardRef<HTMLAnchorElement | HTMLButtonElement, ActionButtonProps>(
    (
        {
            className,
            icon,
            children,
            href,
            size = 's',
            view = 'primary',
            type = 'button',
            iconWrapperClassName,
            disabled,
            loading,
            dataTestId,
            colors = 'default',
            ...rest
        },
        ref,
    ) => {
        const componentRef = useRef<HTMLElement>(null);

        const [focused] = useFocus(componentRef, 'keyboard');

        const { showLoader } = useLoader(!!loading, LOADER_MIN_DISPLAY_INTERVAL);

        const componentProps = {
            className: cn(
                styles.component,
                colorStyles[colors][view],
                styles[size],
                {
                    [styles.focused]: focused,
                    [styles.disabled]: disabled,
                    [styles.loading]: showLoader,
                },
                className,
            ),
            'data-test-id': dataTestId,
        };

        const buttonChildren = (
            <Fragment>
                <span
                    role='img'
                    className={cn(
                        styles.iconWrapper,
                        colorStyles[colors].iconWrapper,
                        styles[size],
                        iconWrapperClassName,
                    )}
                >
                    {showLoader ? <Loader dataTestId='loader' /> : icon}
                </span>
                <span className={styles.label}>{children}</span>
            </Fragment>
        );

        if (href) {
            return (
                <a
                    role='button'
                    ref={mergeRefs([componentRef, ref])}
                    href={href}
                    aria-disabled={disabled || loading}
                    {...componentProps}
                    {...(rest as AnchorHTMLAttributes<HTMLAnchorElement>)}
                >
                    {buttonChildren}
                </a>
            );
        }

        return (
            <button
                ref={mergeRefs([componentRef, ref])}
                // eslint-disable-next-line react/button-has-type
                type={type as ButtonHTMLAttributes<HTMLButtonElement>['type']}
                disabled={disabled || loading}
                {...componentProps}
                {...(rest as ButtonHTMLAttributes<HTMLButtonElement>)}
            >
                {buttonChildren}
            </button>
        );
    },
);
