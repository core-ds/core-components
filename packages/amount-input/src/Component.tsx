import React, { FocusEvent, forwardRef, Fragment, useCallback, useEffect, useState } from 'react';
import cn from 'classnames';

import { Input, InputProps } from '@alfalab/core-components-input';
import { withSuffix } from '@alfalab/core-components-with-suffix';
import { CurrencyCodes } from '@alfalab/data';
import { formatAmount, THINSP } from '@alfalab/utils';

import {
    getAmountValueFromStr,
    getCurrencyCodeWithFormat,
    getFormattedValue,
    getVisiblePlaceholder,
} from './utils';

import defaultColors from './default.module.css';
import styles from './index.module.css';
import invertedColors from './inverted.module.css';

const colorStyles = {
    default: defaultColors,
    inverted: invertedColors,
};

export type AmountInputProps = Omit<InputProps, 'value' | 'onChange' | 'type'> & {
    /**
     * Денежное значение в минорных единицах
     * Значение null - значит не установлено
     */
    value?: string | number | null;

    /**
     * Формат отображения кода валюты
     */
    codeFormat?: 'letter' | 'symbolic';

    /**
     * default - не отображаем копейки, если их значение 0
     * withZeroMinorPart - отображаем копейки, даже если их значение равно 0
     * @default default
     */
    view?: 'default' | 'withZeroMinorPart';

    /**
     * Валюта
     */
    currency?: CurrencyCodes;

    /**
     * Дополнительный закрепленный текст справа от основного значения. (по умолчанию — символ валюты)
     */
    suffix?: string;

    /**
     * Максимальное число знаков до запятой
     * max 15
     */
    integerLength?: number;

    /**
     * Минорные единицы
     */
    minority?: number;

    /**
     * Позволяет вводить только целые значения
     */
    integersOnly?: boolean;

    /**
     * @default - true. Нельзя вводить отрицательные значения.
     * Возможность вводить только положительные значения
     */
    positiveOnly?: boolean;

    /**
     * Жир
     */
    bold?: boolean;

    /**
     * Обработчик события изменения значения
     */
    onChange?: (
        e: React.ChangeEvent<HTMLInputElement>,
        payload: {
            /**
             * Денежное значение в минорных единицах
             * Значение null - значит не установлено
             */
            value: number | null;
            /**
             * Значение инпута
             */
            valueString: string;
        },
    ) => void;

    /**
     * Делает минорную часть полупрозрачной
     */
    transparentMinor?: boolean;
};

/**
 * Инпут, позволяющий закрепить значок валюты
 */
const SuffixInput = withSuffix(Input);

/**
 * Компонент для ввода денежных значений
 */
export const AmountInput = forwardRef<HTMLInputElement, AmountInputProps>(
    (
        {
            view = 'default',
            value = null,
            integerLength: integerLengthProp = 9,
            minority = 100,
            currency = 'RUR',
            suffix = currency,
            codeFormat = 'symbolic',
            placeholder = `0\u2009${
                suffix === currency ? getCurrencyCodeWithFormat(currency, codeFormat) || '' : suffix
            }`,
            integersOnly = false,
            positiveOnly = true,
            bold = true,
            colors = 'default',
            className,
            focusedClassName,
            dataTestId,
            clear = false,
            onChange,
            onClear,
            onBlur,
            breakpoint = 1024,
            onKeyDown,
            transparentMinor = true,
            inputClassName,
            label,
            labelView,
            ...restProps
        },
        ref,
    ) => {
        const integerLength = Math.min(integerLengthProp, 15);

        const getFormattedAmount = useCallback(
            (val: string | number | null) => {
                if (val === '' || val === null || val === '-') return '';

                return formatAmount({
                    value: +val,
                    currency,
                    minority,
                    view,
                    negativeSymbol: 'hyphen-minus',
                }).formatted;
            },
            [currency, minority, view],
        );

        const [inputValue, setInputValue] = useState<string>(() => getFormattedAmount(value));

        const [majorPart, minorPart] = inputValue.split(',');
        const currencyCode = getCurrencyCodeWithFormat(currency, codeFormat);

        useEffect(() => {
            const currentAmountValue = getAmountValueFromStr(inputValue, minority);

            if (currentAmountValue !== value) {
                setInputValue(getFormattedAmount(value));
            }
            // eslint-disable-next-line react-hooks/exhaustive-deps
        }, [value, getFormattedAmount]);

        const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
            const input = e.target;
            let enteredValue = input.value
                .replace(/\s/g, '')
                .replace('.', ',')
                .replace(/[^0-9,-]/g, '');

            if (integersOnly) {
                [enteredValue] = enteredValue.split(',');
            }

            // Эта проверка нужна для того, чтобы обрабатывать значение, переданное в input, длина которого превышает integerLength
            const integer = integersOnly ? enteredValue : enteredValue.split(',')[0];

            if (integer.length > integerLength) {
                enteredValue = enteredValue.slice(0, integerLength);
            }

            // Сокращение минимальной длины мажорной части числа до 0 позволяет ввести "," => "0,"
            const isCorrectEnteredValue = RegExp(
                `(^${positiveOnly ? '' : '-?'}[0-9]{0,${integerLength}}(,([0-9]+)?)?$|^\\s*$)`,
            ).test(enteredValue);

            if (isCorrectEnteredValue) {
                const newFormattedValue = getFormattedValue(enteredValue, currency, minority);

                if (newFormattedValue === inputValue) {
                    const caret = input.selectionStart;

                    window.requestAnimationFrame(() => {
                        input.selectionStart = caret;
                        input.selectionEnd = caret;
                    });
                } else {
                    /**
                     * Поддержка положения каретки
                     * Поскольку при форматировании введенного значения могут появляться символы типа пробела
                     * или запятая - каретка прыгает в конец и ее необходимо ставить в правильное место
                     */

                    // Узнаем длину оригинального инпута с условием обрезания лишних символов

                    const [head, tail] = input.value.split(/\.|,/);
                    let notFormattedEnteredValueLength = head.length;

                    if (tail) {
                        notFormattedEnteredValueLength += 1; // запятая или точка
                        notFormattedEnteredValueLength += tail.slice(
                            0,
                            minority.toString().length - 1,
                        ).length; // символы в минорной части
                    }

                    const diff = newFormattedValue.length - notFormattedEnteredValueLength;
                    const caret = (input.selectionStart as number) + diff;

                    window.requestAnimationFrame(() => {
                        input.selectionStart = caret;
                        input.selectionEnd = caret;
                    });
                }

                setInputValue(newFormattedValue);
                onChange?.(e, {
                    value: getAmountValueFromStr(newFormattedValue, minority),
                    valueString: newFormattedValue,
                });
            } else {
                // Не двигаем каретку когда вставляется невалидный символ
                const caret = (input.selectionStart as number) - 1;

                window.requestAnimationFrame(() => {
                    input.selectionStart = caret;
                    input.selectionEnd = caret;
                });
            }
        };

        const handleKeyDown = (event: React.KeyboardEvent<HTMLInputElement>) => {
            const isModifierKeys = event.ctrlKey || event.altKey || event.metaKey || event.shiftKey;

            // Не двигаем каретку когда вводится невалидный символ
            if (!isModifierKeys && event.key.length === 1 && /[^0-9,.-]/.test(event.key)) {
                event.preventDefault();
            }

            onKeyDown?.(event);
        };

        const handleClear = useCallback(
            (event: React.MouseEvent<HTMLButtonElement>) => {
                setInputValue('');

                if (onClear) {
                    onClear(event);
                }
            },
            [onClear],
        );

        /**
         * Отбросить десятичный разделитель если находится в конце числа
         * 123, => 123
         */
        const dropDecimalSeparator = (event: FocusEvent<HTMLInputElement>) => {
            if (inputValue.endsWith(',')) {
                const pattern = /[,\s]/g; // пробелы и запятые
                const newValue = Number(inputValue.replace(pattern, '')) * minority;
                const formatted = getFormattedAmount(newValue);

                setInputValue(formatted);
                onChange?.(event, {
                    value: newValue,
                    valueString: formatted,
                });
            }
        };

        const handleBlur = (event: FocusEvent<HTMLInputElement>) => {
            if (view === 'withZeroMinorPart') {
                const newValue = getAmountValueFromStr(inputValue, minority);

                if (newValue !== null) {
                    const formatted = getFormattedAmount(newValue);

                    if (formatted !== inputValue) {
                        setInputValue(formatted);
                        onChange?.(event, {
                            value: newValue,
                            valueString: formatted,
                        });
                    }
                }
            }

            if (view === 'default') {
                dropDecimalSeparator(event);
            }

            onBlur?.(event);
        };

        return (
            <div
                className={cn(styles.container, {
                    [styles.bold]: bold,
                    [styles.filled]: Boolean(inputValue),
                })}
            >
                <SuffixInput
                    {...restProps}
                    suffix={
                        <Fragment>
                            {majorPart}

                            <span
                                className={cn({
                                    [colorStyles[colors].minorPartAndCurrency]: transparentMinor,
                                    [colorStyles[colors].disabled]: restProps.disabled,
                                    [colorStyles[colors].readOnly]: restProps.readOnly,
                                })}
                            >
                                {minorPart !== undefined && `,${minorPart}`}
                                {THINSP}
                                {suffix === currency ? currencyCode : suffix}
                            </span>
                        </Fragment>
                    }
                    suffixContainerClassName={styles.suffixContainer}
                    clear={clear}
                    labelView={labelView}
                    label={label}
                    placeholder={getVisiblePlaceholder(placeholder, label, labelView)}
                    value={inputValue}
                    colors={colors}
                    className={cn(styles.component, className)}
                    focusedClassName={focusedClassName}
                    inputClassName={cn(styles.input, inputClassName)}
                    onChange={handleChange}
                    onClear={handleClear}
                    onBlur={handleBlur}
                    onKeyDown={handleKeyDown}
                    inputMode='decimal'
                    pattern={`[${positiveOnly ? '' : '\\-'}0-9\\s\\.,]*`}
                    dataTestId={dataTestId}
                    ref={ref}
                    breakpoint={breakpoint}
                />
            </div>
        );
    },
);

AmountInput.displayName = 'AmountInput';
SuffixInput.displayName = 'SuffixInput';
