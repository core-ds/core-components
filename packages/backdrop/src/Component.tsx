import React, { MouseEvent, ReactNode } from 'react';
import { CSSTransition } from 'react-transition-group';
import { CSSTransitionClassNames } from 'react-transition-group/CSSTransition';
import { TransitionProps } from 'react-transition-group/Transition';
import cn from 'classnames';

import styles from './index.module.css';

export type BackdropProps = Partial<TransitionProps> & {
    /**
     * Прозрачный бэкдроп
     */
    invisible?: boolean;

    /**
     * Управляет видимостью компонента
     */
    open: boolean;

    /**
     * Обработчик клика по бэкдропу
     */
    onClose?: (event: MouseEvent<HTMLElement>) => void;

    /**
     * Дополнительный класс
     */
    className?: string;

    /**
     * Классы анимации
     *
     * http://reactcommunity.org/react-transition-group/css-transition#CSSTransition-prop-classNames
     */
    transitionClassNames?: string | CSSTransitionClassNames;

    /**
     * Идентификатор для систем автоматизированного тестирования
     */
    dataTestId?: string;

    /**
     * Дочерние элементы.
     */
    children?: ReactNode;
};

export const Backdrop: React.FC<BackdropProps> = ({
    className,
    open = false,
    invisible = false,
    timeout = 200,
    children,
    onClose,
    dataTestId,
    transitionClassNames = styles,
    ...restProps
}) => (
        <CSSTransition
            timeout={timeout}
            unmountOnExit={true}
            classNames={transitionClassNames}
            in={open}
            appear={true}
            {...restProps}
        >
            <div
                aria-hidden={true}
                onClick={onClose}
                data-test-id={dataTestId}
                className={cn(styles.backdrop, className, {
                    [styles.invisible]: invisible,
                })}
            >
                {children}
            </div>
        </CSSTransition>
    );
