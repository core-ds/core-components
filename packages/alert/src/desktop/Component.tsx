import React from 'react';

import { PlateDesktop } from '@alfalab/core-components-plate/desktop';

import { type AlertProps, AlertBase } from '../Component';

export const AlertDesktop = (props: AlertProps) => <AlertBase {...props} Plate={PlateDesktop} />;
