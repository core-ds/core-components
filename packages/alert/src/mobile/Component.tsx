import React from 'react';

import { PlateMobile } from '@alfalab/core-components-plate/mobile';

import { type AlertProps, AlertBase } from '../Component';

export const AlertMobile = (props: AlertProps) => <AlertBase {...props} Plate={PlateMobile} />;
