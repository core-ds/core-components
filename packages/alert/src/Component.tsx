import React, { FC } from 'react';
import cn from 'classnames';

import { type Plate, type PlateProps } from '@alfalab/core-components-plate';
import { type PlateDesktop } from '@alfalab/core-components-plate/desktop';
import { type PlateMobile } from '@alfalab/core-components-plate/mobile';

import styles from './index.module.css';

export type AlertProps = Omit<PlateProps, 'foldable' | 'defaultFolded' | 'leftAddons'>;

type AlertPrivateProps = {
    Plate: typeof Plate | typeof PlateDesktop | typeof PlateMobile;
};

export const AlertBase: FC<AlertProps & AlertPrivateProps> = ({
    className,
    title,
    Plate,
    ...restProps
}) => (
    <Plate
        className={cn(styles.component, className)}
        title={title ? <span className={styles.title}>{title}</span> : null}
        rounded={false}
        limitContentWidth={false}
        {...restProps}
    />
);
