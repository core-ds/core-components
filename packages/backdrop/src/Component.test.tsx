import React from 'react';
import { fireEvent, render } from '@testing-library/react';

import { act } from 'react-dom/test-utils';
import { Backdrop } from './Component';

jest.useFakeTimers();

describe('Backdrop', () => {
    const waitForTransition = () =>
        act(() => {
            jest.advanceTimersByTime(300);
        });

    it('should set `data-test-id` attribute', () => {
        const dataTestId = 'test-id';
        const { queryByTestId } = render(<Backdrop open={true} dataTestId={dataTestId} />);

        expect(queryByTestId(dataTestId)).toBeInTheDocument();
    });

    it('should match snapshot', async () => {
        const { queryByTestId, rerender } = render(
            <Backdrop open={true} dataTestId='Backdrop'>
                <span>children</span>
            </Backdrop>,
        );

        // appear
        expect(queryByTestId('Backdrop')).toMatchSnapshot();

        // appear done
        waitForTransition();

        expect(queryByTestId('Backdrop')).toMatchSnapshot();

        // close
        rerender(
            <Backdrop open={false} dataTestId='Backdrop'>
                <span>children</span>
            </Backdrop>,
        );

        // exit
        expect(queryByTestId('Backdrop')).toMatchSnapshot();

        // exit done, unmount
        waitForTransition();

        expect(queryByTestId('Backdrop')).toMatchSnapshot();
    });

    it('should use custom classes', async () => {
        const classNames = {
            appear: 'my-appear',
            appearActive: 'my-active-appear',
            appearDone: 'my-done-appear',
            enter: 'my-enter',
            enterActive: 'my-active-enter',
            enterDone: 'my-done-enter',
            exit: 'my-exit',
            exitActive: 'my-active-exit',
            exitDone: 'my-done-exit',
        };

        const { queryByTestId, rerender } = render(
            <Backdrop
                open={true}
                unmountOnExit={false}
                transitionClassNames={classNames}
                dataTestId='Backdrop'
            />,
        );

        // appear
        expect(queryByTestId('Backdrop')).toHaveClass(classNames.appear, classNames.appearActive);

        // appear done
        waitForTransition();

        expect(queryByTestId('Backdrop')).toHaveClass(classNames.appearDone, classNames.enterDone);

        // close
        rerender(
            <Backdrop
                open={false}
                unmountOnExit={false}
                transitionClassNames={classNames}
                dataTestId='Backdrop'
            />,
        );

        // exit
        expect(queryByTestId('Backdrop')).toHaveClass(classNames.exit, classNames.exitActive);

        // exit done
        waitForTransition();

        expect(queryByTestId('Backdrop')).toHaveClass(classNames.exitDone);

        // reopen
        rerender(
            <Backdrop
                open={true}
                unmountOnExit={false}
                transitionClassNames={classNames}
                dataTestId='Backdrop'
            />,
        );

        // enter
        expect(queryByTestId('Backdrop')).toHaveClass(classNames.enter, classNames.enterActive);

        // enter done
        waitForTransition();

        expect(queryByTestId('Backdrop')).toHaveClass(classNames.enterDone);
    });

    it('should call `onClick` prop', () => {
        const cb = jest.fn();
        const dataTestId = 'test-id';
        const { getByTestId } = render(
            <Backdrop open={true} onClick={cb} dataTestId={dataTestId} />,
        );

        fireEvent.click(getByTestId(dataTestId));

        expect(cb).toBeCalledTimes(1);
    });

    it('should unmount without errors', () => {
        const { unmount } = render(<Backdrop open={true} />);

        expect(unmount).not.toThrowError();
    });
});
