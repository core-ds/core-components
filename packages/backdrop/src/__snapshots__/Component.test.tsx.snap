// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Backdrop should match snapshot 1`] = `
<div
  aria-hidden="true"
  class="backdrop appear appearActive"
  data-test-id="Backdrop"
>
  <span>
    children
  </span>
</div>
`;

exports[`Backdrop should match snapshot 2`] = `
<div
  aria-hidden="true"
  class="backdrop appearDone enterDone"
  data-test-id="Backdrop"
>
  <span>
    children
  </span>
</div>
`;

exports[`Backdrop should match snapshot 3`] = `
<div
  aria-hidden="true"
  class="backdrop exit exitActive"
  data-test-id="Backdrop"
>
  <span>
    children
  </span>
</div>
`;

exports[`Backdrop should match snapshot 4`] = `null`;
